containerz/containerz.proto

```diff
@@ -64,6 +80,22 @@ service Containerz {
   // (i.e. kill) a container.
   rpc StopContainer(StopContainerRequest) returns (StopContainerResponse) {}
 
+  // UpdateContainer updates a running container to the image specified in the
+  // request. By default the operation is synchronous which means that the
+  // request will only return once the container has either been successfully
+  // updated or the update has failed. If the client requests an asynchronous
+  // update then the server must perform all validations (e.g. does the
+  // requested image exist on the system or does the instance name exist) and
+  // return to the client and the update happens asynchronously. If the update
+  // fails, the server must restore the previous version of the container. It
```

Comment:
> the previous version of the container

The previous container is likely to be deleted; it is possible to restore the previous image / launch params, but the container is likely to be re-created. The end result is likely to be identical in most cases, but non-persistent data might be lost (if any).

One potential downside of doing this type of combined operation on the server side is that the rollback result will be invisible to the caller: it will be notified that the new deployment failed, but it won't get a notification whether the server was able to launch the previous version or not.

Reply:
That's correct the previous container instance will most likely be gone and any data within it will be lost unless the data was written to a volume. That volume can be reattached when the upgrade terminates. Generally storing data within the container is considered an anti pattern for exactly this reason.

 Yes in async mode the client would need to check the result of the update but otherwise the result of the upgrade would be communicated back to client.

Reply:
>Yes in async mode the client would need to check the result of the update but otherwise the result of the upgrade would be communicated back to client.

Result of the upgrade - yes, but if the upgrade failed, the last operation ("restoring the previous version") will not produce a result that can be communicated back to the client. I see that you added an additional error code to the latest revision:
```
   // The update failed and the previous version was restored.
    NOT_UPDATED = 4;
```

Which, presumably, means that now the result will be returned only after the rollback is done?

Note that another possible outcome is that the update failed and the previous version was _not_ restored; not sure if you want to add another explicit code for that.

Reply:
That's correct, in synchronous mode if the update fails the rpc should only return after the rollback is done.

I added a error code for a generic failure.